Serialization improvements step 6 (all except wallet/gui)

sipa opened the pull request:
The next step of changes from #10785. 

This:
* Adds support for enum serialization to `CustomUintFormatter`, used in `CAddress` for service flags.
* Merges `BigEndian` into `CustomUintFormatter`, used in `CNetAddr` for port numbers.
* Converts everything (except wallet and gui) to use the new serialization framework.

sipa committed:
Merge BigEndian functionality into CustomUintFormatter

sipa committed:
Convert merkleblock to new serialization

sipa committed:
Convert blockencodings_tests to new serialization

sipa committed:
Convert everything except wallet/qt to new serialization

ryanofsky commented on src/protocol.h:
```diff
@@ -323,23 +314,18 @@ class CAddress : public CService
+            (nVersion >= CADDR_TIME_VERSION && !(s.GetType() & SER_GETHASH))) {
+            READWRITE(obj.nTime);
+        }
+        READWRITE(Using<ConvertFormatter<uint64_t>>(obj.nServices));
```
In commit "Convert everything except wallet/qt to new serialization" (4a185971f495ab89d869a4744d36aa1a8c9aea3f)

I suggested dropping ConvertFormatter in another comment for a few reasons. CustomUintFormatter could be used instead here if extended to support enums:

```diff
diff --git a/src/protocol.h b/src/protocol.h
index c1a21264705..901471e110d 100644
--- a/src/protocol.h
+++ b/src/protocol.h
@@ -324,7 +324,7 @@ public:
             (nVersion >= CADDR_TIME_VERSION && !(s.GetType() & SER_GETHASH))) {
             READWRITE(obj.nTime);
         }
-        READWRITE(Using<ConvertFormatter<uint64_t>>(obj.nServices));
+        READWRITE(Using<CustomUintFormatter<8>>(obj.nServices));
         READWRITEAS(CService, obj);
     }
 
diff --git a/src/serialize.h b/src/serialize.h
index 51d270d7976..718bdea90a6 100644
--- a/src/serialize.h
+++ b/src/serialize.h
@@ -527,14 +527,15 @@ struct CustomUintFormatter
 
     template <typename Stream, typename I> void Unser(Stream& s, I& v)
     {
-        static_assert(std::numeric_limits<I>::max() >= MAX && std::numeric_limits<I>::min() <= 0, "CustomUintFormatter type too small");
+        using U = typename std::conditional<std::is_enum<I>::value, std::underlying_type<I>, std::common_type<I>>::type::type;
+        static_assert(std::numeric_limits<U>::max() >= MAX && std::numeric_limits<U>::min() <= 0, "Assigned type too small");
         uint64_t raw = 0;
         if (BigEndian) {
             s.read(((char*)&raw) + 8 - Bytes, Bytes);
-            v = be64toh(raw);
+            v = static_cast<I>(be64toh(raw));
         } else {
             s.read((char*)&raw, Bytes);
-            v = le64toh(raw);
+            v = static_cast<I>(le64toh(raw));
         }
     }
 };
```

sipa replied:
I've turned this into a separate commit with you as author.

ryanofsky reviewed (commented):
Reviewed most of this (maybe 80%). Just posting feedback so far. Only maybe-blocking thing is question about CAddress::Init below, everything else feel free to ignore.

The last commit here is especially nice. New serialization code is much cleaner than the old code.

ryanofsky commented on src/merkleblock.h:
```diff
@@ -157,17 +161,10 @@ class CMerkleBlock
-        READWRITE(txn);
-    }
-
-private:
```
In commit "Convert merkleblock to new serialization" (25bfbbeb63eca1f1c8852a8a08844069d9e06776)

Maybe better to keep constructor below private and just declare SERIALIZE_METHODS before it, assuming there's a valid reason for the constructor to be private.

sipa replied:
I think this was a mistake; fixed.

ryanofsky reviewed (approved):
Code review ACK 4a185971f495ab89d869a4744d36aa1a8c9aea3f. Really nice improvements in this PR. Left some suggestions and comments, but feel free to ignore. Only thing I would to see answered is question about whether `CAddress::Init` call can be skipped in https://github.com/bitcoin/bitcoin/pull/18317#discussion_r398625457.

ryanofsky committed:
Add SER_READ and SER_WRITE for read/write-dependent statements

Extracted and extended by Pieter Wuille from a comment by Russ
Yanofsky (see
https://github.com/bitcoin/bitcoin/pull/18317#discussion_r398625457).

ryanofsky committed:
Extend CustomUintFormatter to support enums

Extracted by Pieter Wuille from a comment by Russ Yanofsky, see
https://github.com/bitcoin/bitcoin/pull/18317#discussion_r398821936.

sipa commented on src/protocol.h:
```diff
@@ -323,23 +314,18 @@ class CAddress : public CService
+    SERIALIZE_METHODS(CAddress, obj)
     {
-        if (ser_action.ForRead())
-            Init();
```
That's great! I've turned this snippet into a separate commit (with you as author; please let me know if you prefer a different approach).

MarcoFalke replied:
Going forward, the Init should probably be removed and replaced with a slim and clean C++11 initializer

sipa replied:
What do you mean with "C++11 initializer" ?

MarcoFalke replied:
> What do you mean with "C++11 initializer" ?


#19020

sipa replied:
Commented there.

sipa commented on src/serialize.h:
```diff
@@ -540,6 +540,26 @@ struct CustomUintFormatter
 };
 
+/** Formatter that (de)serializes a type as another type it's convertible to/from. */
+template<typename F> struct ConvertFormatter
```
I've taken your approach of merging it into `CustomUintFormatter`.

sipa commented on src/bloom.h:
```diff
@@ -66,15 +66,7 @@ class CBloomFilter
-        READWRITE(nTweak);
-        READWRITE(nFlags);
-    }
+    SERIALIZE_METHODS(CBloomFilter, obj) { READWRITE(obj.vData, obj.nHashFuncs, obj.nTweak, obj.nFlags); }
```
I think that would be pretty hard to do, especially the classname argument to `SERIALIZE_METHODS` is hard to figure out with a regexp. Do you have any concrete ideas?

MarcoFalke replied:
Clang has tooling to do these kind of refactors. Though, I haven't figured out how to run them in a scripted-diff.

sipa commented:
Addressed a number of comments, and included some of @ryanofsky's code suggestions. I've also revamped the merkleblock serializers using the new FOR_READ/FOR_WRITE.

ryanofsky reviewed (approved):
Code review ACK 4eb5643e3538863c9d2ff261f49a9a1b248de243. Changes since last review: new SER_READ/SER_WRITE macros, CustomUintFormatter enum support, updating CPartialMerkleTree to use SER_READ/WRITE and keep constructor private, changing CAddress to call Init() and use CustomUintFormatter

jonatack reviewed (commented):
Cleanup and improvements look very good. Nice collaboration.

ACK 4eb5643e3538863c9d2ff261f49a9a1b248de243

Code review, gcc+clang builds+tests+ran bitcoind at 4eb5643 and rebased on master 65276c7

sipa commented:
Anyone else feel like reviewing this? We're getting close.

practicalswift commented:
Strong concept ACK

New code is much easier to read and reason about.

maflcko reviewed (commented):
ACK 4eb5643e3538863c9d2ff261f49a9a1b248de243 🐜

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 4eb5643e3538863c9d2ff261f49a9a1b248de243 🐜
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUiapgv5AV3Ayc3O2WW2CHMAP08zdDmwuS4qvzMAKT4kAgIdFrVr+iXUXZIwuwQ0
M/HnR17FSJ8FlQ5plt0pyKolQv1fjVugdzJUPd9RjdhXZ3SqYof5arr8ddBi+xJv
qd9XkI8Ur65zVRTOSqWOabWPwGsVxZsztYxwZI8n5yQwWeZ1ANr7O6D/8xEBwggS
tYpFxtTaQ2Wp+REbpgWMYf9t9U2KUW3XFbac2xUf9Ff3hVM4uC+siOGKTf6omaEK
gpO2HT0RzKua6FRgLoTjdczhvPh1RbgLwnMSS7LazgM1rnwTS/fRSsFnr1D+CHqi
fShwPe9+7lOYoC+AzuK4z1RVOfVcLeUFzKRaGQNLXZQJ3LlzIdwg/nhd1xO+qov2
2BF9pbxei6fCfGKz5KCDbJJI9KPcL0K2gfrpldGdF1tZQ7/CbvusEPCIoUe5K6g3
oJ1ttbViqpIGNFRhG0h6JWmu7mOoQm1EuoAjwDz05wJ+H2IthrbRSCYP6BvV00A2
qx5QVuZ8
=A3ac
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `dab97ed81bacef9fec86c29c2502d3a9220b4eae2601a1671bb80dc4cce9dd01  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108dab97ed81bacef9fec86c29c2502d3a9220b4eae2601a1671bb80dc4cce9dd01f0103e5448ea3b3fa3c829f4c71ac7a80a6908fff0101ac55415c90653c96b495989fd5972c308f1045ec436acf008f9eaf7c4976fb5790083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff01043de8d678ce1139069634ab0a296027f08f1045ec436acf0086b6852785fc6a65e0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010a3d018f12f46768f14059b04cca654f808f1045ec436acf00834951f1bec3acb710083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6df010d708b71f50da75ddc16351fe4b82d90008f1045ec436abf008e0dd6bb6c35cba0d0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

sipa committed:
Add comments to CustomUintFormatter

sipa commented on src/serialize.h:
```diff
@@ -527,52 +527,30 @@ struct CustomUintFormatter
- * Use this wrapper around integer types that are stored in memory in native
- * byte order, but serialized in big endian notation. This is only intended
- * to implement serializers that are compatible with existing formats, and
- * its use is not recommended for new data structures.
```
I suspect it was just lost in refactoring. Added a new commit that adds it back.

MarcoFalke commented on src/serialize.h:
```diff
@@ -520,6 +520,15 @@ struct VarIntFormatter
+ * Use the big endian mode for values that are stored in memory in native
+ * byte order, but serialized in big endian notation. This is only intended
+ * to implement serializers that are compatible with existing formats, and
+ * its use is not recommended for new data structures.
```
in commit f9ee0f37c28f604bc82dab502ce229c66ef5b3b9

Not recommend for big endian or anything that can be serialized with this wrapper? For example, I am wondering if there is a recommended way to serialize enums or should new code rather not use enums instead?

sipa replied:
The comment is only intended to apply to big endian serialization. Any suggestions to make that clearer?

If it's confusing, I think it's also fine to just drop the comment. When picking a serialization for a new structure I doubt anyone would intentionally complicate their code by using big endian serialization.

MarcoFalke replied:
> I doubt anyone would intentionally complicate their code by using big endian serialization

Makes sense

ryanofsky replied:
In commit "Add comments to CustomUintFormatter" (f9ee0f37c28f604bc82dab502ce229c66ef5b3b9)

re: "not recommended for new data structures", it'd be nice to say what preferred alternatives are. Compact ints, varints, little endian fixed width ints? I keep forgetting all the int types myself..

sipa replied:
@ryanofsky Maybe the serialization code just shouldn't contain protocol design advice.

MarcoFalke commented:
re-ACK f9ee0f37c2, only change is new documentation commit for CustomUintFormatter 📂

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK f9ee0f37c2, only change is new documentation commit for CustomUintFormatter 📂
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUh+HQv9EdvBDfmaDb6c9KM50spORaYYJlbuXGYi+pomY/sL9zDHQw5DlnO6CsKS
CZ98PFhMSYoTY0jHxLDcQNoJdKnpuS027pJ82pE0lWLzawtSQrCjLk04UkkZwf1h
BvSTfmLOq+071pwCFaSo60n07nIlAlSq7gIGGLq/9UfCM0eAk+ndOWzIxS+lglkR
sO0U6bBK6w8PdpfyrYIJln+db4t7k6FnVeOOJG4TKF3vysvenSN3kknPzch7T7sN
qnVPtist4j8vNAWRxo51BcO42AcGmJM8lK6gqrkXRnn+CqhygxUfQ8oM6Q5UQt6N
mDYRUwvAYTJDwNrPJV378Otgb3Q6ZYF0/+RtIXO44NyhQYZ2mce600UhrrJatXu1
Dbvck/LPkFOkrjucmW4pE0a9xCx+4UCQ7uzPhz51yCXw0fY9fPK7L3OF/XcFTvTT
qNDFffulWh3yjw+0TFbvHSB655J9nabJvIhs9pz8G4FE1+kHNwj0bdjkAaMXkR4y
37zK6NZj
=NdvU
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `ab46775e38a58bfb2a237b81708371ffad21f2374a35d117be61bd7b9e0fd5cf  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108ab46775e38a58bfb2a237b81708371ffad21f2374a35d117be61bd7b9e0fd5cff010afaa031c1cf2dc1fde6b269f3f6b4c9608fff0103f5370fe2553d9b8c99be1c530a9b3e008f120021b00362ccae77f8289a1157f9174dcfae132704506a3347c915b3efdb4595f08f1045ec45707f008bd7ca9946334941c0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0106a9457797f3c728f8123fa6ae820687008f1045ec45707f00825828a4187a58f610083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff01075f5c65caaf71c83e453ab63ff97181508f1045ec45707f00898da2eb82c34fc800083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010d81922116e5ee4b3532eb0a51296c1f908f1045ec45707f00887ac6873423f00010083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6d)

</details>

ryanofsky reviewed (approved):
Code review ACK f9ee0f37c28f604bc82dab502ce229c66ef5b3b9. Just new commit adding comment since last review

jonatack reviewed (commented):
Code review re-ACK f9ee0f37c28f604bc82dab502ce229c6 only change since last review is an additional commit adding Doxygen documentation for `CustomUintFormatter`.
